File: practice_handout/linked_lists/testing_for_cycle.rb
Node = Struct.new(:value, :successor)

List = Struct.new(:head) do
  def cyclic?()
    a, b = self.head, self.head
    while a.successor && b.successor
      a = a.successor
      b = b.successor.successor
      return true if a == b
    end
    false
  end
end

c = Node.new(47, nil)
b = Node.new(23, c)
a = Node.new(1, b)
x = Node.new(3, a)
y = Node.new(2, x)

list = List.new(a)

p list.cyclic?

File: practice_handout/linked_lists/testing_for_cycle.js
var Node = function(value, successor) {
  this.value = value;
  this.successor = successor;
}

var List = function(head) {
  this.head = head;
}

List.prototype.cyclic = function() {
  a = this.head;
  b = this.head;
  while (a.successor && b.successor) {
    a = a.successor;
    b = b.successor.successor;
    if (a === b) {
      return true;
    }
  }
  return false;
}

c = new Node(47)
b = new Node(23,. c)
a = new Node(1, b)
c.successor = a

list = new List(a)

console.log(list.cyclic())